inc/updater/namespace.php

```diff
@@ -1,30 +1,88 @@
 <?php
+/**
+ * Update FAIR packages.
+ *
+ * @package FAIR
+ */
 
 namespace FAIR\Updater;
 
-use Fragen\Git_Updater;
+use FAIR\Packages;
 
-/**
- * Exit if called directly.
- */
-if ( ! defined( 'WPINC' ) ) {
-	die;
-}
+const RELEASE_PACKAGES_CACHE_KEY = 'fair-release-packages';
 
 /**
  * Bootstrap.
  */
 function bootstrap() {
 	add_action( 'init', __NAMESPACE__ . '\\run' );
+	add_filter( 'upgrader_pre_download', __NAMESPACE__ . '\\upgrader_pre_download', 10, 1 );
+}
+
+/**
+ * Add FAIR ReleaseDocument data to cache.
+ *
+ * @param string $did DID.
+ * @return void
+ */
+function add_package_to_release_cache( string $did ) : void {
+	if ( empty( $did ) ) {
+		return;
+	}
+	$releases = wp_cache_get( RELEASE_PACKAGES_CACHE_KEY ) ?: [];
+	$releases[ $did ] = Packages\get_latest_release_from_did( $did );
+	wp_cache_set( RELEASE_PACKAGES_CACHE_KEY, $releases );
+}
+
+/**
+ * Send upgrader_pre_download filter to maybe_add_accept_header().
+ *
+ * @param bool $false Whether to bail without returning the package.
+ *                    Default false.
+ * @return bool
+ */
+function upgrader_pre_download( $false ) : bool {
+	add_filter( 'http_request_args', __NAMESPACE__ . '\\maybe_add_accept_header', 20, 2 );
+	return $false;
+}
+
+/**
+ * Maybe add accept header for release asset package binary.
+ *
+ * ReleaseDocument artifact package content-type will be application/octet-stream.
+ * Only for GitHub release assets.
+ *
+ * @param array  $args Array of http args.
+ * @param string $url  Download URL.
+ *
+ * @return array
+ */
+function maybe_add_accept_header( $args, $url ) : array {
+	$releases = wp_cache_get( RELEASE_PACKAGES_CACHE_KEY ) ?: [];
+
+	if ( ! str_contains( $url, 'api.github.com' ) ) {
```

Comment:
I really don't want to add this exception for a specific source if we can at all avoid it, and I'm pretty certain we can with a properly constructed standard Accept header.

`Accept` indicates data types in preferential order (as indicated by `q=`), so we should be able to use a "universal" Accept header, similar to browsers. Browsers send `text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8` eg

For example, `Accept: application/zip;q=1.0, application/octet-stream;q=0.7, */*;q=0.1` would indicate we accept zips, then binary streams, then anything (in that order).

I'll test this out.

Reply:
Fails in testing.

Reply:
This is caused on the _other_ side of things, where `api.github.com` is incorrectly wanting `Accept: application/json` despite giving a zip response. I've contacted GitHub Support.

It looks like GitHub are also not parsing the `q` parameter correctly, so `*/*;q=0.1` or `*/*;q=0` incorrectly receive priority over `application/octet-stream`. This is noncompliant with the HTTP spec afaik.